src/CodeXMLParse.cpp

```diff
@@ -46,6 +46,15 @@ void AnnotateOpref(ANNOTATOR_PARAMS)
 	annotation.type = R_CODE_ANNOTATION_TYPE_OFFSET;
 	annotation.offset.offset = op->getAddr().getOffset();
 }
+void AnnotateFunctionName(ANNOTATOR_PARAMS){
+	auto func_name = node.child_value();
```

Comment:
(not related to the code here, making this comment to start a nested discussion)
> It would be more reliable to use address of the function directly instead of the name. While the name can be evaluated to obtain address it's messy and you need the correct name. There is function name, real name, flag name, demangled name. The address of function referenced at position of code would also have other uses.

I didn't think that renaming using the name would get messy. But initially, I wanted to have the address in the function name annotations, just so that it might become useful later. But unfortunately, we don't have this information in XML. So I am not sure how I could do it @karliss  @thestr4ng3r

Reply:
https://github.com/radareorg/radare2/blob/master/libr/include/r_anal.h#L277-L280
Does this name always refer to the 'function name' and not real name/flag name/demangled name.

Reply:
The address can be clearly seen in the xml.

Reply:
@karliss I am unable to find it till now. We have the address of the current function in the XML(and I knew we could get that).   But how can I find the addresses of all the functions in the decompiled code from the XML?

Reply:
@NirmalManoj Can you post an example xml here for reference?

Reply:
I had a call with Nirmal looking at the xml, Ghidra C++ API and Opref annotator as example for accessing Ghidra C++ API. He should be good for a while.

Reply:
@karliss @thestr4ng3r  I was trying this all day and it seems I need more help. This is what I am trying finally (not working code)
```
        PcodeOp *op = opit->second;
	uintb temp_off = 0;
	for(int i = 0; i < op->numInput(); i++){
		const Varnode *vn = op->getIn(i);
		if (vn->getSpace()->getType() == IPTR_FSPEC){
			temp_off = vn->getOffset();
			// temp_off++;
		}
	}
	annotation.function_name.offset = temp_off;
	out->push_back(annotation);
```
It does identify functions correctly with `IPTR_FSPEC`, but `vn->getOffset()` is not returning the correct offset. I want to find a way to get the offset from the `Varnode` and I believe there is a way to do that from `vn`. But I couldn't figure that out.

Reply:
a) How did you came to conclusion that IPTR_FSPEC is the correct address space kind?
b) did you to dump all the varnodes you get with their properties: type, whatever offset returns, the id displayed in xml, anything else potentially interesting by printing it to terminal?
c) The offsets in xml looked reasonable, take a look a t how xml is printed

Reply:
a) In `space.hh` in Ghidra,
```
/// \brief Fundemental address space types
///
/// Every address space must be one of the following core types
enum spacetype {
  IPTR_CONSTANT = 0,	       ///< Special space to represent constants
  IPTR_PROCESSOR = 1,	       ///< Normal spaces modelled by processor
  IPTR_SPACEBASE = 2,	       ///< addresses = offsets off of base register
  IPTR_INTERNAL = 3,	       ///< Internally managed temporary space
  IPTR_FSPEC = 4,	       ///< Special internal FuncCallSpecs reference
  IPTR_IOP = 5,                ///< Special internal PcodeOp reference
  IPTR_JOIN = 6		       ///< Special virtual space to represent split variables
};
```
From this and the code that I have read, I guessed `IPTR_FSPEC` will be the only `spacetype` for function calls. So I tried it.
c) I was trying to find a way to get the offsets without extracting it from the XML. I did check saveXml functions (but not ruling out the possibility that I won't get anything from it in future)
b) I haven't done all of it. I will try it now.

Reply:
I used a debugger to see the offsets of all the `varnodes` inside `ctx->func.vbank.loc_tree`. I will attach the binary below. It has 25 nodes under `varnodes` tag in XML output and it displays the offsets of all those correctly except those that give us function offsets. 
( Attached image shows the location that I am talking about clearly)
![Screenshot from 2020-07-07 18-48-29](https://user-images.githubusercontent.com/18501167/86787397-abdeac80-c082-11ea-873f-76c64049efc1.png) ![Screenshot from 2020-07-07 18-53-32](https://user-images.githubusercontent.com/18501167/86787885-30312f80-c083-11ea-90ed-555d33950c86.png)

It will be great if someone with experience of Ghidra API could help me in understanding it more. I have spent a lot of time in the past two days trying to figure this out. @karliss @thestr4ng3r

Reply:
I think I have finally found a way to extract offsets reliably.

Reply:
@karliss @thestr4ng3r This PR is ready for more review.